clang/lib/CodeGen/CGBlocks.cpp

```diff
@@ -2764,6 +2794,9 @@ void CodeGenFunction::emitByrefStructureInit(const AutoVarEmission &emission) {
     auto layoutInfo = CGM.getObjCRuntime().BuildByrefLayout(CGM, type);
     storeHeaderField(layoutInfo, getPointerSize(), "byref.layout");
   }
+
+  if (emission.NeedsInitOnHeap)
+    emitByrefInitOnHeap(pointer);
```

Comment:
So, this is not exception-safe, right?  If the initialization throws, we have no way to release the byref object we're allocating here.

Now, it looks like the [blocks runtime](https://github.com/llvm/llvm-project/blob/0a93e9f2e264472f1c8c3b5d8726aa43a02ecace/compiler-rt/lib/BlocksRuntime/runtime.c#L368) is itself not exception-safe, so maybe we can hand-wave this?  `_Block_byref_assign_copy ` allocates, calls the copy operation, and makes no effort to deallocate on a throw.  On the other hand, the byref "copy" operation is typically a call to a move constructor, so it is a lot less likely in practice that it'll throw than that the original initialization will.

The only way I can think of to handle this would be to actually put a "successfully initialized" flag in the byref payload and then no-op the destructor if it isn't set.  That's implementable but pretty awful.  So maybe we should just acknowledge it's not exception-safe and suggest that the blocks runtime add new APIs to support this properly.  What do you think?

Reply:
Argh, you're right.

In principle the `BLOCK_HAS_COPY_DISPOSE` flag could be used for that purpose, instead of creating a separate field in the payload.  It could start out unset (making the runtime skip the call to `byref_dispose`), and then be set after initialization.

But the initialization might send the block capturing the object to another thread.  If initialization then throws, I assume that nobody will subsequently _call_ the block, or if they do, the block won't actually try to use the object.  That's straightforward UB.  But it doesn't seem like it should be UB just to send the block.  So there might be concurrent calls to `_Block_object_dispose`.  Not only that, 'our' `_Block_object_dispose` call (the one that would hypothetically be added on the cleanup path to fix the memory leak) might not be the one to drop the last reference.

Therefore, `flags` would have to be modified atomically, particularly because the same word is also used for the reference count.  And even with an atomic operation, you would still have theoretical UB in the blocks runtime when it reads `flags` non-atomically.  Though, not any more UB than it's already incurring thanks to potential concurrent refcount modifications...  But what if an alternate block runtime tries to cache `flags`?  The only alternate runtime I know of is GNUstep's and it looks like [it doesn't cache it](https://github.com/gnustep/libobjc2/blob/master/blocks_runtime.m#L199), but this still feels very dubious.

So then there's the alternative you suggested of just acknowledging that this isn't exception-safe (specifically, that it leaks memory if initialization throws).  But that makes this whole patch feel like an awkward middle ground.  If self-capturing `__block` variables are not going to be supported properly (pending new APIs), then maybe they shouldn't be supported at all and should just be an error (pending new APIs).

But if new APIs are added, teaching Clang how to recognize when the target OS supports them would be quite a lot of work for what is ultimately an edge case.

I'm going to see how ugly it ends up being to add a "successfully initialized" flag.

Reply:
Updated to add the flag.  It's not all that bad.

In fact, if, as I suggested in my last comment, we want to support the byref lingering after unwinding, then there *must* be some kind of runtime state – to track whether whoever drops the last reference needs to run the destructor.  It would be more natural for the Blocks runtime to track that state rather than relying on compiler-generated code, but the state has to live somewhere.

If, on the other hand, we don't want to support the byref lingering after unwinding, then I could go back to the idea of mutating the existing `flags` field.  In that scenario, the existence of runtime state would be unnecessary, but the implementation would be relatively unobtrusive.

Reply:
Ping.  Any thoughts?  I know this is a mess :)

If both options are too messy, there's also the option of making these self-captures a hard error, which would not break backwards compatibility since they never worked properly.

Reply:
Ping.